pkg/gardenlet/controller/shoot/care/health.go

```diff
@@ -666,6 +680,30 @@ func (h *Health) CheckClusterNodes(
 	return nil, nil
 }
 
+func CheckNodeAgentLease(nodeList *corev1.NodeList, leaseList *coordinationv1.LeaseList, clock clock.Clock) error {
+
+	for _, node := range nodeList.Items {
+		var nodeLease *coordinationv1.Lease = nil
+		for _, lease := range leaseList.Items {
```

Comment:
Instead of having this nested loop, maybe you build a `map[string]coordinationv1.Lease` first by iterating over the `leaseList`. Key of the map would be the node name (can be derived from the lease name).
This way, you can access the `Lease` object in `O(1)` when iterating over the nodes here.

Reply:
After this is done, you need to add another early exit (`return nil`) in case this map is empty. This can only happen when there is no `Lease`. This can happen when the `Shoot` was not yet reconciled after the `UseGardenerNodeAgent` feature gate has been turned on.